src/or/connection_edge.c

```diff
@@ -446,6 +454,172 @@ connection_edge_end(edge_connection_t *conn, uint8_t reason)
+   * data. */
+  if (conn->base_.state != AP_CONN_STATE_RESOLVE_WAIT) {
+    // How many more data cells can arrive on this id?
+    half_conn->data_pending = conn->deliver_window;
+  }
+
+  smartlist_add(circ->half_streams, half_conn);
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so,
+ * check if it still has data cells pending, and decrement that
+ * window if so.
+ *
+ * Return 1 if the data window was not empty.
+ * Return 0 otherwise.
+ */
+int
+connection_half_edge_is_valid_data(smartlist_t *half_conns,
+                                   streamid_t stream_id)
+{
+  if (!half_conns)
+    return 0;
+
+  SMARTLIST_FOREACH_BEGIN(half_conns, half_edge_t *, half_conn) {
+    if (half_conn->stream_id == stream_id) {
+      if (half_conn->data_pending > 0) {
+        half_conn->data_pending--;
+        return 1;
+      } else {
+        return 0;
+      }
+    }
+  } SMARTLIST_FOREACH_END(half_conn);
+
+  return 0;
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so,
+ * check if it still has a connected cell pending, and decrement
+ * that window if so.
+ *
+ * Return 1 if the connected window was not empty.
+ * Return 0 otherwise.
+ */
+int
+connection_half_edge_is_valid_connected(smartlist_t *half_conns,
+                                        streamid_t stream_id)
+{
+  if (!half_conns)
+    return 0;
+
+  SMARTLIST_FOREACH_BEGIN(half_conns, half_edge_t *, half_conn) {
+    if (half_conn->stream_id == stream_id) {
+      if (half_conn->connected_pending > 0) {
+        half_conn->connected_pending--;
+        return 1;
+      } else {
+        return 0;
+      }
+    }
+  } SMARTLIST_FOREACH_END(half_conn);
+
+  return 0;
+}
+
+/**
+ * Streams that were used to send a RESOLVE cell are closed
+ * when they get the RESOLVED, without an end. So treat
+ * a RESOLVED just like an end, and remove from the list.
+ */
+int
+connection_half_edge_is_valid_resolved(smartlist_t *half_conns,
+                                       streamid_t stream_id)
+{
+  return connection_half_edge_is_valid_end(half_conns, stream_id);
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so, remove
+ * it from the list. No other data should come after the END cell.
+ *
+ * Return 1 if stream_id was in half-closed state.
+ * Return 0 otherwise.
+ */
+int
+connection_half_edge_is_valid_end(smartlist_t *half_conns,
+                                  streamid_t stream_id)
+{
+  half_edge_t *found_half = NULL;
+  if (!half_conns)
+    return 0;
+
+  SMARTLIST_FOREACH_BEGIN(half_conns, half_edge_t *, half_conn) {
+    if (half_conn->stream_id == stream_id) {
+      found_half = half_conn;
+    }
+  } SMARTLIST_FOREACH_END(half_conn);
+
+  if (found_half) {
+    smartlist_remove(half_conns, found_half);
+    tor_free(found_half);
+    return 1;
+  }
+
+  return 0;
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so,
+ * check if it still has sendme cells pending, and decrement that
+ * window if so.
```

Comment:
I factored out the search code as per above. I did the regrouping of similar functions in e6bf1cfa1b62554243c3fe066af86c43e90ac912, since it also seems like a good idea.

Reply:
Now I want github with --color-moved